internal/wrapper.go

```diff
@@ -75,22 +63,76 @@ func RunKubeStateMetricsWrapper(opts *options.Options) {
 		})
 		viper.WatchConfig()
 	}
+
+	if opts.CustomResourceConfigFile != "" {
+		WatchConfig(opts.CustomResourceConfigFile, func(e fsnotify.Event) {
+			klog.Infof("Changes detected: %s\n", e.Name)
+			cancel()
+			// Wait for the ports to be released.
+			<-time.After(3 * time.Second)
+			ctx, cancel = context.WithCancel(context.Background())
+			go KSMRunOrDie(ctx)
+		})
+	}
 	klog.Infoln("Starting kube-state-metrics")
 	KSMRunOrDie(ctx)
 	select {}
 }
 
-func resolveCustomResourceConfig(opts *options.Options) (customresourcestate.ConfigDecoder, bool) {
-	if s := opts.CustomResourceConfig; s != "" {
-		return yaml.NewDecoder(strings.NewReader(s)), true
-	}
-	if file := opts.CustomResourceConfigFile; file != "" {
-		f, err := os.Open(filepath.Clean(file))
+// WatchConfig is taken and adapted from https://github.com/spf13/viper/blob/b89e554a96abde447ad13a26dcc59fd00375e555/viper.go#L429 MIT-Licensed
+func WatchConfig(filename string, onConfigChange func(fsnotify.Event)) {
```

Comment:
Couldn't we just create a Viper config for the custom resource config and reuse the original function?

Reply:
I'm not sure if Viper is able to handle multiple configs at the same time and independently.

Reply:
We could create new viper instances with `viper.New()` and stop using the global viper config for the general config.

A benefit to reusing viper directly here is that we would be able to use `viper.ReadInConfig()` and `viper.ReadInConfig()` like we do for the general config.